design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+
+Note that users can also execute `DROP <STREAM/TABLE>` requests via `executeDdl(...)` or `executDml(...)` above.
+
+#### `TERMINATE <queryId>`
+
+```
+  CompletableFuture<TerminateQueryResponse> terminatePersistentQuery(String queryId);
+```
+where `TerminateQueryResponse` is again the same as `DdlResponse`/`DmlResponse`/`CreateSourceResponse`/`InsertIntoResponse`/`DropSourceResponse` above.
+
+The method name `terminatePersistenQuery(...)` is to distinguish from `terminatePushQuery(...)` below.
+
+#### Connectors
+
+```
+  CompletableFuture<ConnectorInfo> createSourceConnector(String name, Map<String, String> properties);
+
+  CompletableFuture<ConnectorInfo> createSinkConnector(String name, Map<String, String> properties);
+
+  CompletableFuture<Void> dropConnector(String name);
+```
+where `ConnectorInfo` is from an Apache Kafka module ([link](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/ConnectorInfo.java)).
+
+Or would we rather not have a dependency on Apache Kafka in the ksqlDB client interfaces?
+
+### Admin operations
+
+#### `SHOW TOPICS`
+
+```
+  CompletableFuture<List<TopicInfo>> listTopics();
+```
+with
+```
+public interface TopicInfo {
+
+  String getName();
+
+  int getPartitions();
+  
+  List<Integer> getReplicasPerPartition();
+
+}
+```
+
+#### `SHOW <STREAMS/TABLES>`
+
+```
+  CompletableFuture<List<StreamInfo>> listStreams();
+
+  CompletableFuture<List<TableInfo>> listTables();
+```
+with
+```
+public interface StreamInfo {
+
+  String getName();
+
+  String getTopic();
+
+  String getFormat();
+  
+}
+```
+and
+```
+public interface TableInfo {
+
+  String getName();
+
+  String getTopic();
+
+  String getFormat();
+
+  boolean isWindowed();
+  
+}
+```
+
+I'm not sure whether it makes more sense for `StreamInfo#getFormat()` and `TableInfo#getFormat()` to return a string or an enum value.
+The latter would make it easier for the user to know the possible values, but we'd have to keep the list up to date and would also sacrifice forward compatibility.
+
+#### `SHOW QUERIES`
+
+```
+  CompletableFuture<List<QueryInfo>> listQueries();
+```
+with
+```
+public interface QueryInfo {
+
+  boolean isPersistentQuery();
+
+  boolean isPushQuery();
+
+  /**
+   * Query ID, used for control operations such as terminating the query
+   */
+  String getId();
+
+  String getSql();
+
+  /**
+   * Name of sink, for a persistent query. Else, null.
+   */
+  String getSink();
+
+  /**
+   * Name of sink topic, for a persistent query. Else, null.
+   */
+  String getSinkTopic();
+  
+  /**
+   * Map of query state ("RUNNING", "ERROR", or "UNRESPONSIVE") to the number of ksqlDB servers
+   * on which this query is in each state.
+   */
+  Map<String, Integer> getServerStatusCounts();
+
+}
+```
+
+Again, not sure whether it makes more sense to represent query states as strings, or to introduce an Enum and have `getServerStatusCounts()` return an EnumMap instead.
```

Comment:
not sure I see how this can be used programatically from a client... can we punt on introducing this method at all?

Reply:
This is the only client method that exposes query status, so if a user wanted to check the health of their queries (either periodically, or if they suspect something is wrong) this is what they'd use. Are you saying checking query status isn't a typically client use case? (Not necessarily disagreeing, just trying to understand.)

Reply:
hmm, I'd be in favor of exposing less and then adding functionality as users request it (in line with Tim's less is more mantra)

Reply:
I also agree with the idea of initially exposing a minimal amount of environment/server metadata. If we expose as much as possible before users have expressed a need for it, we risk having to make breaking changes in the future if the way users eventually need to use it ends up being a bit different than we assumed.

Reply:
less is more, repeat ad infinitum ;)